### PR TITLE
Move creating a SyncConfiguration to SyncUser

### PR DESCRIPTION
**Pull request by cmelchior**
As discussed, moving to Query-based Sync as the default needs to be less breaking, so this PR moves creating the SyncConfiguration to SyncUser and reintroduces the old methods but marks them as deprecated. All tests have been converted as well.

The interesting files to review are `SyncUser` and `SyncConfiguration`

### CODE REVIEW

**Comment by cmelchior** [comment]
Ready for review.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,9 +1,22 @@
...
+
 ### Breaking Changes
 
-* [ObjectServer] Partial sync is now the default mode of synchronization. `SyncConfiguration.Builder.partialRealm()` has been replaced by `SyncConfiguration.Builder.fullSynchronization()`.
-* [ObjectServer] `SyncConfiguration.isPartialRealm()` has been replaced by `SyncConfiguration.isFullySynchronizedRealm()`.
+* [ObjectServer] `SyncConfiguration.automatic()` has been replaced by `SyncUser.getDefaultConfiguration()`.
+
+### Deprecated
+
+* [ObjectServer] `new SyncConfiguration.Builder(user, url)` has been deprecated in favour of `SyncUser.createConfiguration(url)`. NOTE: The Builder constructors sync mode is still full synchronization while the new `createConfiguration()` uses query-based sync as the default mode.
```
Comment (nhachicha) [inline CHANGELOG.md]: > The Builder constructors sync mode is still full synchronization

this need rephrasing & including the old deprecated ctor as an example

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,9 +1,22 @@
...
+* [ObjectServer] `SyncConfiguration.Builder.partialRealm()` has been deprecated in favor of `SyncConfiguration.Builder.fullSynchronization()`.
```
Comment (nhachicha) [inline CHANGELOG.md]: maybe rephrase to:

Query-Based Sync is now the default, `SyncConfiguration.Builder.partialRealm()` has been deprecated, use  `SyncConfiguration.Builder.fullSynchronization()` if you want full synchronisation instead.

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncConfigurationTests.java [hunk 3]
```diff
@@ -75,20 +75,6 @@ public void tearDown() {
         }
     }
 
-    @Test
-    public void user_invalidUserThrows() {
```
Comment (nhachicha) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncConfigurationTests.java]: Why removing this test?

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 4]
```diff
@@ -450,6 +384,21 @@ boolean isSyncConfiguration() {
         return sessionStopPolicy;
     }
 
+    /**
+     * Whether this configuration is for a query-based Realm.
+     * <p>
+     * Query-based synchronization allows a synchronized Realm to be opened in such a way that
+     * only objects queried by the user are synchronized to the device.
+     *
+     * @return {@code true} to open a partial synchronization Realm {@code false} otherwise.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: to open a query-based Realm

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 5]
```diff
@@ -450,6 +384,21 @@ boolean isSyncConfiguration() {
...
+     * @see Builder#partialRealm() for more details.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: We refer to a deprecated doc?

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 6]
```diff
@@ -521,9 +471,12 @@ public boolean isFullySynchronizedRealm() {
          *            assume the file is located on the same server returned by {@link SyncUser#getAuthenticationUrl()}.
          *
          * @see SyncUser#isValid()
+         * @deprecated Use {@link SyncUser#createConfiguration(String)} instead.
          */
+        @Deprecated
         public Builder(SyncUser user, String uri) {
             this(BaseRealm.applicationContext, user, uri);
+            fullSynchronization(); //
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: remove empty comment

File: realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java [hunk 7]
```diff
@@ -957,15 +910,25 @@ public Builder waitForInitialRemoteData() {
             return this;
         }
 
+        /**
+         * Setting this will open a partially synchronized Realm.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncConfiguration.java]: partially (query-based)

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 8]
```diff
@@ -212,6 +213,79 @@ public SyncUser run() throws ObjectServerError {
...
+     * <p>
+     * The {@code /~/} will automatically be replaced with the user ID when creating the {@link SyncConfiguration}.
+     * <p>
+     * Moreover, the URI defines the local location on disk. The location of a synchronized Realm file is
+     * {@code /data/data/<packageName>/files/realm-object-server/<user-id>/<last-path-segment>}, but this behavior
+     * can be overwritten using {@link SyncConfiguration.Builder#name(String)} and {@link SyncConfiguration.Builder#directory(File)}.
+     * <p>
+     * Many Android devices are using FAT32 file systems. FAT32 file systems have a limitation that
+     * file names cannot be longer than 255 characters. Moreover, the entire URI should not exceed 256 characters.
+     * If file name and underlying path are too long to handle for FAT32, a shorter unique name will be generated.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: If the file

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 9]
```diff
@@ -212,6 +213,79 @@ public SyncUser run() throws ObjectServerError {
...
+     */
+    public SyncConfiguration.Builder createConfiguration(String uri) {
+        if (!isValid()) {
+            throw new IllegalStateException("Configurations can only be created from valid users");
+        }
+        return new SyncConfiguration.Builder(this, uri).partialRealm();
+    }
+
+    /**
+     * Returns the default configuration for this user. The default configurations points to the
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: The default configuration

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 10]
```diff
@@ -212,6 +213,79 @@ public SyncUser run() throws ObjectServerError {
...
+     * default query-based Realm on the server the user authenticated.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: authenticated against

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,9 +1,22 @@
 ## 6.0.0 (YYYY-MM-DD)
 
+The feature previously named Partial Sync is now called Query-Based Sync and is now the default mode when synchronizing Realms.
+This has impacted a number of API's. See below for the details.
+
 ### Breaking Changes
 
-* [ObjectServer] Partial sync is now the default mode of synchronization. `SyncConfiguration.Builder.partialRealm()` has been replaced by `SyncConfiguration.Builder.fullSynchronization()`.
-* [ObjectServer] `SyncConfiguration.isPartialRealm()` has been replaced by `SyncConfiguration.isFullySynchronizedRealm()`.
+* [ObjectServer] `SyncConfiguration.automatic()` has been replaced by `SyncUser.getDefaultConfiguration()`.
```
Comment (bmunkholm) [inline CHANGELOG.md]: Why not just deprecated so we avoid making a breaking change and a major version update?
Related: This should no goto next-major branch or be 6.0 then.

Comment (cmelchior) [reply CHANGELOG.md]: Yes, that was the plan. Looks like an oversigt when reverting the breaking changes.
